### ISSUE TITLE
Update to Ruby2.2, ActiveRecord4 and Rspec3

### ISSUE THREAD

**Issue by topofocus** [ib-ruby/ib-ruby#95]
Hi alltogether,
The application ist somewhat outdated.
I forked the current master to do some refreshments (see title...).

Current Status: Testsuite passes for all model-specs.
I included Guard-support to the testsuite and defined some Models with GirlFactory

Next-step: ActiveRecord4-Support, which is now thread-save and a major improvement for using ib-ruby in production

forked project: topofocus/ib-ruby

everybody is invited for critism and support

**Comment by arvicco** [comment]
Yes, at this point current maintainers do not have time to update the project. If you're interested to take more active role in it, you're welcome. Invitation to join ib-ruby sent.

**Comment by topofocus** [comment]
Thanks.
Should we create a dev-branch to concentrate the update/upgrade there?

**Comment by arvicco** [comment]
Yes, I think it would be advisable.

From:  topofocus <notifications@github.com>
Reply-To:  ib-ruby/ib-ruby
<reply+00009a8509833fb15b48a1a19d3e8dd0711b8168f03f48fb92cf0000000110c10ce29
2a169ce0327efb7@reply.github.com>
Date:  Sunday, January 4, 2015 at 06:14
To:  ib-ruby/ib-ruby <ib-ruby@noreply.github.com>
Cc:  Ar Vicco <arvicco@gmail.com>
Subject:  Re: [ib-ruby] Update to Ruby2.2, ActiveRecord4 and Rspec3 (#95)


Thanks.
Should we create a dev-branch to concentrate the update/upgrade there?

‹
Reply to this email directly or view it on GitHub
<https://github.com/ib-ruby/ib-ruby/issues/95#issuecomment-68634178> .

**Comment by topofocus** [comment]
Hi, I created a branch AR4. I could'nd figure out, how to put my forked project into the branch. 
How is this done?

**Comment by arvicco** [comment]
When creating a Pull Request in your Github fork, there is Edit button on top which allows you to choose which branch of the upstream repo you'd like to contribute to.

**Comment by topofocus** [comment]
Thancs arvicco.
If no rails-integration is needed, the AR4-branch provides a working-solution for ruby2.2-projects.

**Comment by jumper444** [comment]
topofocus,
Confused by your last comment,  You saying the AR4 brance has the rails code stripped out (which was a discussion previously by others) or that if you simply avoid any rails code/integration/calls or such that the package should work fine?

**Comment by jumper444** [comment]
And could you specify which ruby 2.2 you are using in which environment?
I previously has a 1.9/1.8 install issue only to discover arvicco was actually using Jruby while I was using regular ruby on windows.  This created additional actions/issues on my part which were already build into arvicco's production environment.

**Comment by topofocus** [comment]
Hi,
I'm using an Ubuntu environment for development & production. In Production the stable branch works well with rvm/ Ruby 2.0 and 2.1. In dev I switched to the latest 2.2.0 realease.

In the AR4-Branch the Code itself is untouched (yet). 
I updated the Test-Suite and verified, that they pass in the changed environment.
To my findings, its save to use Ruby2.2 even in the stable branch.

In my view, Rails-Intergration adds unnesessery complexity. With the thread-save active-record 4 two independend projects working on the same Database are preferable. 
Therefor I try to obtain a package which relays messages to the tws and puts their results into a database, 

The Gemfile/gemspec in the AR4-Branch is updated accodingly.

Thus – by cloning the AR4-Branch you get a much smaller Installation with an usable Test-Suite (Rspec3/Guard) 
If anyone is interested, I can add the AR4-Frame to integrate the Gem into a ruby-project.
Basics are here:   https://github.com/topofocus/ActiveRecord-without-Rails

**Comment by ricpruss** [comment]
AR4 is still pretty broken.  As it is trying to use Active Record alone it does not have the rake migrations working, I wrote some changes to Rakefile to get it working but then you start to run into all the problems in tws_reader.rb, which has spelling mistakes, under_con_id to con_id mismatch in the test and I gave up shortly after that.

**Comment by topofocus** [comment]
hi Richard,
can you post your changes to the rakefile? 
Which test(s) failed?

**Comment by jumper444** [comment]
Could you guys answer a question that I have.  When I review the various scripting level IB implementations around the web (there are some for javascript/nodejs/python, etc), none of them seem to incorporate additional frameworks or activerecord or whatnot - at least not by default or by requirement.  They just do exactly what you expect...they act to enable the IB API..and that's ALL they do..  Everything else is up to the coder.  Sorta the "unix way"...do one task and do it well. (Then build up with different tasks working together.).

So why does this IB framework need to be intertwined with all kinds of other stuff (rails, activerecord, etc.).  Why can't (or doesn't it) just "do the IB API"?  

(I'm a programmer, but not by profession for many years and no I haven't used rails or activerecord yet....So maybe the reason would be self obvious if I did.)

**Comment by jumper444** [comment]
And if I may ask a followup related. When I look at, for example, the two main javascript API implementations (node-ib, node-ibapi-addon). Each of them say "this is a WRAPPER for [either] the IB C+ API or the Java API".  It alludes them them not as much doing programming of an extension as, instead, just wrapping and gluing together the work that IB is already putting out.

If this interpretation is correct (maybe not) then could ib-ruby just wrap code already coming from IB  (just translating to ruby) instead of being a rewrite or new version?  Could IB's Java implementation simple be translated/linked to and used inside Jruby?

What am I missing in my comment?
Is ib-ruby already doing a wrap?
Was ib-ruby designed in a different way (with the rails and active record stuff) preventing this?
Or, perhaps, IB on ruby simply doesn't have enough interest/users so it lags some?
THanks.

**Comment by jumper444** [comment]
So this actually leads to a third question.  If the javascript IB API modules are (seemingly) pure and up to date, what about using THOSE modules from ruby ( https://github.com/sstephenson/execjs ) instead of ib-ruby.  Or is this just asking for a nightmare?

(I'm not using rails or DBs or object storage or whatever. I'm using command line, text input ruby scripts to automate order entry, order kill, order update, etc.  I know what that stuff is and maybe I need it in the future, but simple API connection is all I need for now.)

**Comment by topofocus** [comment]
Hi Jumper,
if you compare the examples of ib-ruby to those of other scripting-language-api-solutions, you find almost identical approaches for »simple tasks«, as fireing one order or fetching market-data. If things get more complex, the scripting-language itself is important.
Node.js has its advantages as python or go has. If I look to recent developments in ruby, I feel very comfortable with this language. The core of IB-ruby is build with ruby 1.6 and the stable branch officially supports not even ruby 2.0. That's why I started to work on some features.

To my knowlege, only ib-ruby has the advantage of supporting structural data (active-model) from the beginning.  This translates ib-messages directly into ruby-objects which can live in a database for further processing. Basicly ib-ruby provides an environment to translate a ruby-command
    
    ib.send_message (...)
    
to an api-call and to handle recieved ib-messages in a ruby way.  This I would call an »simple wrapper«.
Later this approach was extended to enable ib-ruby to work as an engine in a rails-framework.

I do agree, that this might be an overkill. 
Therfore I am suggesting a slight downsizing of the project. 
Personally, I wont miss the active-model features, which enable high-level-validations and further data-processing with active-support-methods. These benefits come cheap. If nessesary, this can easily extended to a complete database-backed-API-solution.

Further, if things  get more sophisticated, ib-ruby (in the ar4-branch) supports multithreading and thus accesses the full power of the ib-api.  

In conclusion, I do agree that we have to downsize the project and to focus on stability and readable code. I started with IB::Contracts, encapsulated some api-calls into the method »update-contract« (which can be used without any database).  

Next will be a default skeleton to manage api-calls (a gateway-module). 

We have to upgrade ib-ruby to the latest ib-api-Version, which should be manageable due to the updated testing facilities. However, the benefits are minor. Therfor I focus on improvements using ruby2.2 and ar4.

**Comment by ricpruss** [comment]
In practice you can easily run and test ib-ruby without the active record part.  If you run rspec with -rdb you actually test the database parts but otherwise it should not.

I do not agree that Rails is overkill for me this totally adds value as it makes the basic things I need to do very straightforward, but what we may want to do is split the two layers into two projects. The db integration can consume the messaging layer as a gem.   I believe this may be good as it separates the dependent external forces on the project.  IB-Ruby has to deal with different versions of messaging from the TWS API’s moving along and then different needs from the Rails migrations and changes, and I guess both need to deal with the many changes in the ruby lang.


On 14 February 2015 at 9:38:56 pm, topofocus (notifications@github.com) wrote:
Hi Jumper,
if you compare the examples of ib-ruby to those of other scripting-language-api-solutions, you find almost identical approaches for »simple tasks«, as fireing one order or fetching market-data. If things get more complex, the scripting-language itself is important.
Node.js has its advantages as python or go has. If I look to recent developments in ruby, I feel very comfortable with this language. The core of IB-ruby is build with ruby 1.6 and the stable branch officially supports not even ruby 2.0. That's why I started to work on some features.

To my knowlege, only ib-ruby has the advantage of supporting structural data (active-model) from the beginning. This translates ib-messages directly into ruby-objects which can live in a database for further processing. Basicly ib-ruby provides an environment to translate a ruby-command

ib.send_message (...)

to an api-call and to handle recieved ib-messages in a ruby way. This I would call an »simple wrapper«.
Later this approach was extended to enable ib-ruby to work as an engine in a rails-framework.

I do agree, that this might be an overkill.
Therfore I am suggesting a slight downsizing of the project.
Personally, I wont miss the active-model features, which enable high-level-validations and further data-processing with active-support-methods. These benefits come cheap. If nessesary, this can easily extended to a complete database-backed-API-solution.

Further, if things get more sophisticated, ib-ruby (in the ar4-branch) supports multithreading and thus accesses the full power of the ib-api.

In conclusion, I do agree that we have to downsize the project and to focus on stability and readable code. I started with IB::Contracts, encapsulated some api-calls into the method »update-contract« (which can be used without any database).

Next will be a default skeleton to manage api-calls (a gateway-module).

We have to upgrade ib-ruby to the latest ib-api-Version, which should be manageable due to the updated testing facilities. However, the benefits are minor. Therfor I focus on improvements using ruby2.2 and ar4.

**Comment by arvicco** [comment]
Initially, ib-ruby wasn't using AR, just ActiveModel. AR and Rails integration came later. 

The problem with decoupling ib-ruby from AR and Rails integration stems from the fact that you need to define the type of business object very early on. This is what happens in https://github.com/ib-ruby/ib-ruby/blob/master/lib/ib.rb. And message handlers in https://github.com/ib-ruby/ib-ruby/tree/master/lib/ib/messages create and use these business objects extensively. There is no way around it, so far as I can see.

In practice, ib-ruby could be operated in one of 3 modes:
1) Lightweight ActiveModels, no DB or Rails
2) AR based DB-backed business objects, no Rails
3) AR based DB-backed business objects with Rails

You need to be sure that pretty much the whole test suite pass in all 3 modes (Rails has additional tests as well). So, it's not easy to break off a messaging layer into a separate gem. I considered this, but could not find a good way to do it, and still make sure all 3 modes are adequately tested. It may be possible, but won't be easy, for sure.

Another approach would be "back to basics" where you drop AR/Rails integration and create basic messaging gem with ActiveModels. Then, build AR integration gem on top of it (that should test basic messaging all over again, with new business object type). And then, Rails integration gem on top (that should test basic messaging all over again in Rails environment, and add some Rails-specific tests as well). In this case you need to replicate the spec suit in every gem which is not very DRY and problematic.

Unfortunately, I cannot contribute to dev effort currently due to other full-time commitments.

**Comment by topofocus** [comment]
Let's try to leave no option behind and focus on an up-to-date implementation.
The current ib-ruby lacks a staightforward support of FA-Accounts

In my lab-version (https://github.com/topofocus/ib-ruby) today I uploaded a suggestion.

It implements an Account-Model where (tabelless or db-backed) active advisors and users are defined, as well as a advisor-user-hierachy (db-backed).
Then an IB::Alert class is introduced, which handles the response to :Alert - Messages.

This leads to a third feature: IB::Gateway.
A frame to manage Connections and a basic skeleton for the interaction with the TWS (i.e. through :Alerts ). A future version of Gateway should handle reconnects and other interruptions of the connection to the TWS with bookkeeping of active-accounts and unfilled orders.
Suggestions are welcome.
Check the latest commit at  https://github.com/topofocus/ib-ruby

**Comment by jm3** [comment]
This issue is now a year old :cry: Perhaps the approach of, "leave no option behind" was too ambitious, and we can now leave some options behind? :)

**Comment by topofocus** [comment]
Hi,
the IB::Gateway is just working fine. Several thin-clients  operating in remote-locations are working in 24/7-mode since more then 12 month now. 
The question »lets leave some options behind« is justified.

The Gateway uses the lightweight, non-persistent IB::Model data-representation and runs everywhere. I too struggle with the complexity which was incorporated during the effort, to include a rails-support. 
As the hype thought rails is fading, maybe we can concentrate on the essentials – to provide a reliable and easy to use gateway from ruby-scripts to the TWS-API.

**Comment by jm3** [comment]
I have no idea what your comment is trying to say.
